utils/ixdtf/src/parsers/mod.rs

```diff
@@ -0,0 +1,327 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! The parser module contains the implementation details for `IXDTFParser` and `ISODurationParser`
+
+use crate::{ParserError, ParserResult};
+
+extern crate alloc;
+
+use alloc::{string::String, vec::Vec};
+
+use self::duration::parse_duration;
+use datetime::DateTimeFlags;
+use grammar::is_annotation_open;
+use records::{DateRecord, DurationParseRecord, IsoParseRecord, TimeZone};
+
+pub mod records;
+
+mod annotations;
+pub(crate) mod datetime;
+pub(crate) mod duration;
+mod grammar;
+mod time;
+pub(crate) mod time_zone;
+
+#[cfg(test)]
+mod tests;
+
+/// `assert_syntax!` is a parser specific utility macro for asserting a syntax test, and returning the
+/// the provided provided error if the assertion fails.
+#[macro_export]
+macro_rules! assert_syntax {
+    ($cond:expr, $err:ident $(,)?) => {
+        if !$cond {
+            return Err(ParserError::$err);
+        }
+    };
+}
+
+/// `IxdtfParser` is the primary parser implementation of `ixdtf`.
+///
+/// This parser provides various options for parsing date/time strings with the extended notation
+/// laid out in [sedate's IXDTF][ixdtf-draft] along with other variations laid out in the [`Temporal`][temporal-proposal].
+///
+/// [ixdtf-draft]: https://datatracker.ietf.org/doc/draft-ietf-sedate-datetime-extended/
+/// [temporal-proposal]: https://tc39.es/proposal-temporal/
+#[derive(Debug)]
+pub struct IxdtfParser {
+    cursor: Cursor,
+}
+
+impl IxdtfParser {
+    /// Creates a new `IXDTFParser` from a provided `&str`.
+    pub fn new(target: &str) -> Self {
+        Self {
+            cursor: Cursor::new(target),
+        }
+    }
+
+    /// Parses the source as a [DateTime string][temporal-dt].
+    ///
+    /// This is the baseline parser where the TimeRecord, UTCOffset, and all annotations are optional.
+    ///
+    /// [temporal-dt]: https://tc39.es/proposal-temporal/#prod-TemporalDateTimeString
+    pub fn parse_date_time(&mut self) -> ParserResult<IsoParseRecord> {
+        datetime::parse_annotated_date_time(DateTimeFlags::empty(), &mut self.cursor)
+    }
```

Comment:
Discussion: I think the code that runs the parser differently depending on the Temporal type should stay in the `temporal_rs` crate. It would be reasonable to have flags that control the parser behavior, but how to interpret the return value should be up to the client.

How much do we need these parse flags? I feel like the IXDTF parser can not take any flags and just always return a struct with a bunch of optional fields, plus a map/vec at the end for extra annotations (beyond time zone and calendar).

Basically, the `ixdtf` crate should do one thing only: take a string that is expected to be in IXDTF syntax and parse it into a Rust data structure. Then the consumers of the crate can decide what to do with that structured data.